### PR TITLE
Templates & errors for taxonomies

### PR DESCRIPTION
**Pull request by stejskalleos**
Fix for two issues:
- Show errors in the UI for missing type and taxonomies
- Ignore taxonomies for snippets, unify it with behavior in UI

**TODO**
This must be done for report templates, job templates, etc. But first I'd like to get agreement on the approach before editing other areas and tests.

### CODE REVIEW

File: app/views/provisioning_templates/_alerts.html.erb [hunk 1]
```diff
@@ -1,3 +1,10 @@
 <% if @template.locked? -%>
   <%= locked_warning(::Template.find(params[:id])) %>
 <% end -%>
+
+<% if @template.errors.any? -%>
+<%= alert :header => _("Unable to save"),
+          :class  => 'alert-danger base in fade',
+          :text   => @template.errors.map { |e| ("<li>#{e.attribute}: #{e.message}</li>") }.join.html_safe
```
Comment (ShimShtein) [inline app/views/provisioning_templates/_alerts.html.erb:8]: Worth using some specific methods:
```
[13] pry(main)> hh.errors.messages
=> {:"interfaces.mac"=>["can't be blank"],
 :"interfaces.domain"=>["can't be blank"],
 :"interfaces.domain_id"=>["can't find domain with this id"],
 :operatingsystem_id=>["can't be blank"],
 :organization_id=>["can't be blank"],
 :location_id=>["can't be blank"],
 :architecture_id=>["can't be blank"]}
[14] pry(main)> hh.errors.full_messages
=> ["Mac can't be blank",
 "Domain can't be blank",
 "Interfaces.domain can't find domain with this id",
 "Operatingsystem can't be blank",
 "Organization can't be blank",
 "Location can't be blank",
 "Architecture can't be blank"]
```

This example is for host object, but template object would behave the same.

**Review by ShimShtein — CHANGES_REQUESTED** [review]
Before we can continue I have a couple of questions:
1. What use case are we talking about: view/modify the snippet or provisioning/using the snippet during the render?
2. What as the original need for this PR?

**Comment by stejskalleos** [comment]
There are two issues fixed in PR:
1. Snippets (according to UI) don't need to be scoped by the org/loc.
2. When **non-admin** user without `assign_organization` permission tries to create a snippet, they won't get any message, just edit page without any alert.

**Comment by ShimShtein** [comment]
@ares I need a second opinion on this one. It looks OK to me, but I am not that well-versed in taxonomies.

**Comment by ares** [comment]
> Snippets (according to UI) don't need to be scoped by the org/loc.

What leads you to this? I see I can assign orgs and locs to the snippets just like any other templates.

However I also realize, one can't create a snippet in a different scope (location) that would have the same name. That's general problem with the our tenancy model. Given the snippets are often used for extension points with `snippet_if_exists`, it makes sense to me to make them always shared.

Now what is a bit concerning is the behavior for existing users of the multitenancy. Snippets could contain sensitive information, such as tokens, private keys or what not. If we simply share the snippets, all orgs will have access to this data. We'd have to be very clear in upgrade instructions.

Given how broken the multitenancy is today, I think we can't make it much worse. At the same time, it would be good to have some longer term plan we converge to.

For this PR specifically I'd like to
1) understand why we think UI does not need snippets to be scoped
2) have a plan or be prepared on how we communicate the change

**Comment by stejskalleos** [comment]
> What leads you to this? I see I can assign orgs and locs to the snippets just like any other templates.

In UI, if you check the snippet kind, the form field with organizations and locations disappears. The truth is that it doesn't mean the snippet is not taxable anymore, but the UI behavior suggests it.

> Given how broken the multitenancy is today, I think we can't make it much worse. At the same time, it would be good to have some longer term plan we converge to.

We had this discussion in the past, but AFAIK there hasn't been any agreement.

> Given how broken the multitenancy is today, I think we can't make it much worse. 

What I could do is to split the PR into two, one for the error message, and the second for the taxonomy issue.

**Pull request closed without merging**